### DISCUSSION TITLE
manager/asdf: support multiple tool versions per line

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#36233] [Suggest an Idea]
### Describe the proposed change(s).

Add support for multiple tool versions per line.

Example test today limited to one: https://github.com/renovatebot/renovate/blob/c8fceec593a184cc7b858bb1186dd52c6a1206ed/lib/modules/manager/asdf/extract.spec.ts#L31-L42

Related discussion: #24968

**Comment by Suraj-kumar00** [comment]
Hi @rarkins, How are you doing?
Can I pick this up? And can you please help to understand it more so that I solve this issue?

**Comment by rarkins** [comment]
Please ask in https://github.com/renovatebot/renovate/discussions/24968 if anything there is unclear

**Comment by MPV** [comment]
@Suraj-kumar00 Any plans/ideas on how to solve this?

**Comment by Suraj-kumar00** [comment]
Hi @MPV I want to be honest with you,actually I haven't understood the problem clearly and you haven't replied so i forgot to do this how I'm busy till 14 december because my exams are going on. Like if it's not require to solve this issue fast I can take it after my exams and can you please help me to understand this?

**Comment by rarkins** [comment]
Converted to a discussion.

@MPV I've realized that the specs for this are a little incomplete. It would be easy to write code to get this:

```
    it('only captures the first version', () => {
      const res = extractPackageFile('nodejs 16.16.0 16.15.1');
      expect(res).toEqual({
        deps: [
          {
            currentValue: '16.16.0',
            datasource: 'node-version',
            depName: 'node',
          },
          {
            currentValue: '16.15.1',
            datasource: 'node-version',
            depName: 'node',
          },
        ],
      });
    });
```

But the problem with this is that both of these would get major and minor updates together, so you'd need further config to selectively disable this.

I think we need logic like this:
- If more than one major version is found in the list then disable major updates for lower ones
- If more than one minor version is found with the same major then apply only patch updates for the lower ones

Even this is not perfect though. This issue is getting old, and Node 16 is now EOL. Hard to tell how these lines should have evolved over time. e.g. now 22.16.0 and 22.15.1?

Note: this is not node.js specific or even asdf-specific, but we may as well solve it here first.
